File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
const longestValidParentheses = (s) => {
  const stack = [], index = [], intervals = [];
  for (let i = 0; i < s.length; ++i) {
    const b = s[i];
    if (b === '(') {
      stack.push(b);
      index.push(i);
    } else if (b === ')' && stack.at(-1) === '(') {
      stack.pop();
      const at = index.pop();
      while (intervals.length && intervals.at(-1)[0] > at) {
        intervals.pop();
      }
      if (intervals.length && intervals.at(-1)[1] === at - 1) {
        intervals.at(-1)[1] = i;
      } else {
        intervals.push([at, i]);
      }
    } else {
      stack.slice(0, i);
      index.slice(0, i);
    }
  }

  let max = 0;
  for ([from, to] of intervals) {
    if (max < to - from + 1) max = to - from + 1;
  }
  return max;
};

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
class Solution:
  def longestValidParentheses(self, s: str) -> int:
    stack = []
    index = []
    intervals = []
    for i, ch in enumerate(s):
      if ch == '(':
        stack.append(ch)
        index.append(i)
      elif stack and stack[-1] == '(':
        stack.pop()
        j = index.pop()
        while intervals and intervals[-1][0] > j:
          intervals.pop()
        if intervals and intervals[-1][1] == j - 1:
          intervals[-1][1] = i
        else:
          intervals.append([j, i])
    ans = 0
    for j, i in intervals:
      ans = max(ans, i - j + 1)
    return ans
          